File: benchmarks/handwritten/ackermann.js
const iterations = parseInt(process.argv[2]) || 100;
const inputM = parseInt(process.argv[3]) || 3;
const inputN = parseInt(process.argv[4]) || 7;

function ackermann(m, n) {
  if (m === 0) return n + 1;
  if (n === 0) return ackermann(m - 1, 1);
  return ackermann(m - 1, ackermann(m, n - 1));
}

// Warmup
ackermann(inputM, inputN);

const start = performance.now();
let result;
for (let i = 0; i < iterations; i++) {
  result = ackermann(inputM, inputN);
}
const end = performance.now();

process.stderr.write(`result=${result}\n`);
console.log(Math.round((end - start) * 1000));

File: benchmarks/handwritten/ackermann.cpp
#include <chrono>
#include <cstdlib>
#include <iostream>

long long ackermann(long long m, long long n) {
  if (m == 0)
    return n + 1;
  if (n == 0)
    return ackermann(m - 1, 1);
  return ackermann(m - 1, ackermann(m, n - 1));
}

int main(int argc, char *argv[]) {
  int iterations = argc > 1 ? std::atoi(argv[1]) : 1000;
  int input_m = argc > 2 ? std::atoi(argv[2]) : 3;
  int input_n = argc > 3 ? std::atoi(argv[3]) : 7;

  // Warmup
  ackermann(input_m, input_n);

  auto start = std::chrono::high_resolution_clock::now();
  volatile long long result = 0;
  for (int i = 0; i < iterations; i++) {
    result = ackermann(input_m, input_n);
  }
  auto end = std::chrono::high_resolution_clock::now();

  auto time_us =
      std::chrono::duration_cast<std::chrono::microseconds>(end - start)
          .count();
  std::cerr << "result=" << result << std::endl;
  std::cout << time_us << std::endl;
  return 0;
}
